File: scala/binary/src/main/scala/Binary.scala
case class Binary(str: String) {
  def toDecimal = str.foldLeft[Option[Int]](Some(0)) {
    case (Some(acc), '0') => Some(acc * 2)
    case (Some(acc), '1') => Some(acc * 2 + 1)
    case _ => None
  }.getOrElse(0)
}

File: fsharp/binary/Binary.fs
module Binary

type Binary(input: string) =

    let isValid char = 
        match char with
        | '0' | '1' -> true
        | _  -> false

    let charToDecimal char = (int)char - (int)'0'

    member this.toDecimal() = 
        let chars = input.ToCharArray()
        if Array.forall isValid chars then Array.fold (fun acc c -> acc * 2 + charToDecimal c) 0 chars
        else 0